neo4j/src/main/java/org/apache/metamodel/neo4j/Neo4jDataContext.java

```diff
@@ -0,0 +1,336 @@
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * DataContext implementation for Neo4j
+ */
+public class Neo4jDataContext extends QueryPostprocessDataContext implements UpdateableDataContext,
+        DocumentSourceProvider {
+
+    public static final Logger logger = LoggerFactory.getLogger(Neo4jDataContext.class);
+
+    public static final String SCHEMA_NAME = "neo4j";
+
+    public static final int DEFAULT_PORT = 7474;
+
+    public static final String RELATIONSHIP_PREFIX = "rel_";
+
+    public static final String RELATIONSHIP_COLUMN_SEPARATOR = "#";
+
+    private final SimpleTableDef[] _tableDefs;
+
+    private final Neo4jRequestWrapper _requestWrapper;
+
+    private final HttpHost _httpHost;
+
+    public Neo4jDataContext(String hostname, int port, String username, String password, SimpleTableDef... tableDefs) {
+        _httpHost = new HttpHost(hostname, port);
+        final CloseableHttpClient httpClient = HttpClientBuilder.create().build();
+        _requestWrapper = new Neo4jRequestWrapper(httpClient, _httpHost, username, password);
+        _tableDefs = tableDefs;
+    }
+
+    public Neo4jDataContext(String hostname, int port, String username, String password) {
+        _httpHost = new HttpHost(hostname, port);
+        final CloseableHttpClient httpClient = HttpClientBuilder.create().build();
+        _requestWrapper = new Neo4jRequestWrapper(httpClient, _httpHost, username, password);
+        _tableDefs = detectTableDefs();
+    }
+
+    public Neo4jDataContext(String hostname, int port, CloseableHttpClient httpClient) {
+        _httpHost = new HttpHost(hostname, port);
+        _requestWrapper = new Neo4jRequestWrapper(httpClient, _httpHost);
+        _tableDefs = detectTableDefs();
+    }
+
+    public Neo4jDataContext(String hostname, int port, CloseableHttpClient httpClient, SimpleTableDef... tableDefs) {
+        _httpHost = new HttpHost(hostname, port);
+        _requestWrapper = new Neo4jRequestWrapper(httpClient, _httpHost);
+        _tableDefs = tableDefs;
+    }
+
+    @Override
+    protected String getDefaultSchemaName() throws MetaModelException {
+        return SCHEMA_NAME;
+    }
+
+    @Override
+    protected Schema getMainSchema() throws MetaModelException {
+        MutableSchema schema = new MutableSchema(getMainSchemaName());
+        for (SimpleTableDef tableDef : _tableDefs) {
+            MutableTable table = tableDef.toTable().setSchema(schema);
+            schema.addTable(table);
+        }
+        return schema;
+    }
+
+    @Override
+    protected String getMainSchemaName() throws MetaModelException {
+        return SCHEMA_NAME;
+    }
+
+    public SimpleTableDef[] detectTableDefs() {
+        List<SimpleTableDef> tableDefs = new ArrayList<SimpleTableDef>();
+
+        String labelsJsonString = _requestWrapper.executeRestRequest(new HttpGet("/db/data/labels"));
```

Comment:
Maybe we should put the service root (/db/data) in some constant or even a field. It seems from the docs [1] that while this is the default URL it might be different in certain environments. And at least also that the result of it will provide the URLs for other services.

[1] http://neo4j.com/docs/stable/rest-api-service-root.html